apps/studio/src/components/ResourceSelector/ResourceSelector.tsx

```diff
@@ -1,157 +1,220 @@
+import { useQueryParse } from "~/hooks/useQueryParse"
+import { sitePageSchema } from "~/pages/sites/[siteId]"
 import { trpc } from "~/utils/trpc"
 import { ResourceItem } from "./ResourceItem"
 
+interface ResourceSelectorProps {
+  queryFn:
+    | typeof trpc.resource.getChildrenOf.useInfiniteQuery
+    | typeof trpc.resource.getFolderChildrenOf.useInfiniteQuery
+  onChange: (resourceId: string) => void
+  selectedResourceId?: string
+  existingResource?: PendingMoveResource | null
+}
+
+const generatePermalinkPrefix = (parents: PendingMoveResource[]) => {
+  return parents.map((parent) => parent.permalink).join("/")
+}
+
 const SuspensableResourceSelector = ({
-  siteId,
-  selectedResourceId,
+  queryFn,
   onChange,
-  isDisabledFn,
+  selectedResourceId,
+  existingResource,
 }: ResourceSelectorProps) => {
-  const [ancestryStack] = trpc.resource.getAncestryOf.useSuspenseQuery({
-    siteId,
-    resourceId: selectedResourceId,
-  })
-  const [parentIdStack, setParentIdStack] = useState<string[]>(
-    ancestryStack.map((item) => item.id),
+  // NOTE: This is the stack of user's navigation through the resource tree
+  // NOTE: We should always start the stack from `/` (root)
+  // so that the user will see a full overview of their site structure
+  const [resourceStack, setResourceStack] = useState<PendingMoveResource[]>([])
+  const [isResourceHighlighted, setIsResourceHighlighted] =
+    useState<boolean>(true)
+  const { siteId } = useQueryParse(sitePageSchema)
+
+  const moveDest = resourceStack[resourceStack.length - 1]
+  const parentDest = resourceStack[resourceStack.length - 2]
+  const curResourceId = moveDest?.resourceId
+  const { data, fetchNextPage, hasNextPage, isFetchingNextPage } = queryFn(
+    {
+      resourceId:
+        (isResourceHighlighted
+          ? parentDest?.resourceId
+          : moveDest?.resourceId) ?? null,
+      siteId: String(siteId),
+      limit: 25,
+    },
+    {
+      getNextPageParam: (lastPage) => lastPage.nextOffset,
+    },
   )
-  const currResourceId = parentIdStack[parentIdStack.length - 1] ?? null
+  const ancestryStack: PendingMoveResource[] = trpc.resource.getAncestryWithSelf
+    .useSuspenseQuery({
+      siteId: String(siteId),
+      resourceId: selectedResourceId,
+    })[0]
+    .map((resource) => ({ ...resource, resourceId: resource.id }))
+
+  useEffect(() => {
+    if (
+      ancestryStack.length <= 0 ||
+      JSON.stringify(ancestryStack) === JSON.stringify(resourceStack)
+    ) {
+      return
+    }
+    setResourceStack(ancestryStack)
+  }, [])
 
-  const [data, { fetchNextPage, hasNextPage, isFetchingNextPage }] =
-    trpc.resource.getChildrenOf.useSuspenseInfiniteQuery(
-      {
-        resourceId: currResourceId,
-        siteId,
-        limit: 25,
-      },
-      {
-        getNextPageParam: (lastPage) => lastPage.nextOffset,
-      },
-    )
+  useEffect(() => {
+    if (curResourceId) {
+      onChange(curResourceId)
+    }
+  }, [curResourceId])
```

Comment:
clarification needed - i don't quite understand this section here, i'd expect to see teh `onChange` on a button as that's the trigger for the change rather than on `useEffect`, as that implies we're syncing state somewhere

Reply:
in some cases, the parent component that uses `ResourceSelector` will need `onChange` (props passed in) to trigger state change